akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala

```diff
@@ -0,0 +1,224 @@
+    if (value.isInstanceOf[ORSet[_]] && values.contains(key))
+      throw new IllegalArgumentException(
+        "`ORMap.put` must not be used to replace an existing `ORSet` " +
+          "value, because important history can be lost when replacing the `ORSet` and " +
+          "undesired effects of merging will occur. Use `ORMultiMap` or `ORMap.updated` instead.")
+    else
+      new ORMap(keys.add(node, key), values.updated(key, value))
+
+  /**
+   * Scala API: Replace a value by applying the `modify` function on the existing value.
+   *
+   * If there is no current value for the `key` the `initial` value will be
+   * passed to the `modify` function.
+   */
+  def updated(node: Cluster, key: String, initial: A)(modify: A ⇒ A): ORMap[A] =
+    updated(node.selfUniqueAddress, key, initial)(modify)
+
+  /**
+   * Java API: Replace a value by applying the `modify` function on the existing value.
+   *
+   * If there is no current value for the `key` the `initial` value will be
+   * passed to the `modify` function.
+   */
+  def updated(node: Cluster, key: String, initial: A, modify: akka.japi.Function[A, A]): ORMap[A] =
+    updated(node, key, initial)(value ⇒ modify.apply(value))
+
+  /**
+   * INTERNAL API
+   */
+  private[akka] def updated(node: UniqueAddress, key: String, initial: A)(modify: A ⇒ A): ORMap[A] = {
+    val newValue = values.get(key) match {
+      case Some(old) ⇒ modify(old)
+      case _         ⇒ modify(initial)
+    }
+    new ORMap(keys.add(node, key), values.updated(key, newValue))
+  }
+
+  /**
+   * Removes an entry from the map.
+   * Note that if there is a conflicting update on another node the entry will
+   * not be removed after merge.
+   */
+  def -(key: String)(implicit node: Cluster): ORMap[A] = remove(node, key)
+
+  /**
+   * Removes an entry from the map.
+   * Note that if there is a conflicting update on another node the entry will
+   * not be removed after merge.
+   */
+  def remove(node: Cluster, key: String): ORMap[A] = remove(node.selfUniqueAddress, key)
+
+  /**
+   * INTERNAL API
+   */
+  private[akka] def remove(node: UniqueAddress, key: String): ORMap[A] = {
+    new ORMap(keys.remove(node, key), values - key)
+  }
+
+  override def merge(that: ORMap[A]): ORMap[A] = {
+    val mergedKeys = keys.merge(that.keys)
+    var mergedValues = Map.empty[String, A]
+    mergedKeys.elementsMap.keysIterator.foreach {
+      case key: String ⇒
+        (this.values.get(key), that.values.get(key)) match {
+          case (Some(thisValue), Some(thatValue)) ⇒
+            if (thisValue.getClass != thatValue.getClass) {
+              val errMsg = s"Wrong type for merging [$key] in [${getClass.getName}], existing type " +
+                s"[${thisValue.getClass.getName}], got [${thatValue.getClass.getName}]"
+              throw new IllegalArgumentException(errMsg)
+            }
+            val mergedValue = thisValue.merge(thatValue.asInstanceOf[thisValue.T]).asInstanceOf[A]
```

Comment:
why is the type equality not statically known? it would seem that having the same `A` implies that IAE will not ever be thrown and these casts should be unnecessary …